runners/apex/src/main/java/org/apache/beam/runners/apex/ApexYarnLauncher.java

```diff
@@ -0,0 +1,379 @@
+        return true;
+      }
+      @Override
+      public void shutdown(ShutdownMode arg0) throws LauncherException {
+        // TODO: interaction with child process
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  /**
+   * From the current classpath, find the jar files that need to be deployed
+   * with the application to run on YARN. Hadoop dependencies are provided
+   * through the Hadoop installation and the application should not bundle them
+   * to avoid conflicts. This is done by removing the Hadoop compile
+   * dependencies (transitively) by parsing the Maven dependency tree.
+   *
+   * @return list of jar files to ship
+   * @throws IOException when dependency information cannot be read
+   */
+  public static List<File> getYarnDeployDependencies() throws IOException {
+    InputStream dependencyTree = ApexRunner.class.getResourceAsStream("/dependency-tree");
+    BufferedReader br = new BufferedReader(new InputStreamReader(dependencyTree));
+    String line = null;
+    List<String> excludes = new ArrayList<>();
+    int excludeLevel = Integer.MAX_VALUE;
+    while ((line = br.readLine()) != null) {
+      for (int i = 0; i < line.length(); i++) {
+        char c = line.charAt(i);
+        if (Character.isLetter(c)) {
+          if (i > excludeLevel) {
+            excludes.add(line.substring(i));
+          } else {
+            if (line.substring(i).startsWith("org.apache.hadoop")) {
+              excludeLevel = i;
+              excludes.add(line.substring(i));
+            } else {
+              excludeLevel = Integer.MAX_VALUE;
+            }
+          }
+          break;
+        }
+      }
+    }
+    br.close();
+
+    Set<String> excludeJarFileNames = Sets.newHashSet();
+    for (String exclude : excludes) {
+      String[] mvnc = exclude.split(":");
+      String fileName = mvnc[1] + "-";
+      if (mvnc.length == 6) {
+        fileName += mvnc[4] + "-" + mvnc[3]; // with classifier
+      } else {
+        fileName += mvnc[3];
+      }
+      fileName += ".jar";
+      excludeJarFileNames.add(fileName);
+    }
+
+    ClassLoader classLoader = ApexYarnLauncher.class.getClassLoader();
+    URL[] urls = ((URLClassLoader) classLoader).getURLs();
+    List<File> dependencyJars = new ArrayList<>();
+    for (int i = 0; i < urls.length; i++) {
+      File f = new File(urls[i].getFile());
+      // dependencies can also be directories in the build reactor,
+      // the Apex client will automatically create jar files for those.
+      if (f.exists() && !excludeJarFileNames.contains(f.getName())) {
+          dependencyJars.add(f);
+      }
+    }
+    return dependencyJars;
+  }
+
+  /**
+   * Create a jar file from the given directory.
+   * @param dir source directory
+   * @param jarFile jar file name
+   * @throws IOException when file cannot be created
+   */
+  public static void createJar(File dir, File jarFile) throws IOException {
```

Comment:
The functionality is available as part of apex-common.jar. Please see https://github.com/apache/apex-core/pull/415